crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1748,8 +1748,37 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_arguments(arguments);
         self.infer_expression(func);
 
-        // TODO resolve to return type of `func`, if its a callable type
-        Type::Unknown
+        match func.as_ref() {
+            name_expr @ ast::Expr::Name(..) => {
+                let name_expr_id = name_expr.scoped_ast_id(self.db, self.scope);
+                let Some(named_type) = self.types.expressions.get(&name_expr_id) else {
+                    return Type::Unbound;
+                };
+
+                match named_type {
+                    Type::Function(function_type) => {
+                        function_type.returns(self.db).unwrap_or(Type::Unknown)
+                    }
+
+                    // TODO: handle class constructors
+                    // TODO: handle classes which implement the Callable protocol
+                    Type::Class(_class_type) => Type::Unknown,
+
+                    // `Any` is callable, and its return type is also `Any`.
+                    Type::Any => Type::Any,
+
+                    // TODO: union and intersection types, if they reduce to `Callable`?
+                    Type::Union(_) => todo!(),
+                    Type::Intersection(_) => todo!(),
+
+                    _ => Type::Unknown,
```

Comment:
There aren't a ton of examples of this in the code yet, because we only recently gained the ability to emit diagnostics -- but we do have that ability now, and I think we should start actually emitting them where it is correct to do so. So here we should emit a diagnostic with the message `{type} is not callable`. (We could also add a TODO for that instead if you find it adds a lot of complexity for some reason, but I don't think it should?)

Reply:
Ah, interesting – from previous discussions with @AlexWaygood I was under the (mistaken! And I don’t blame Alex for this) impression that we deferred *all* handling of diagnostics to later in the pipeline. Will update. 👍🏼

Reply:
I think it's not a priority for us to emit comprehensive diagnosis for all typing errors right now, but I agree with @carljm that it makes sense to start emitting diagnostics where it makes sense to do so 👍 it helps us see how things all fit together into a bigger picture